File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [s, p] = input[0].split(" ").map(Number)
const dna = input[1].split("")
const [a, c, g, t] = input[2].split(" ").map(Number)
const currentChar = {
  A: 0,
  C: 0,
  G: 0,
  T: 0,
}
const check = () => {
  if (
    currentChar["A"] >= a &&
    currentChar["C"] >= c &&
    currentChar["G"] >= g &&
    currentChar["T"] >= t
  ) {
    return 1
  }
  return 0
}
let answer = 0
for (let i = 0; i < p; i++) {
  currentChar[dna[i]] += 1
}
answer += check()

for (let i = 1; i < s - p + 1; i++) {
  currentChar[dna[i - 1]] -= 1
  currentChar[dna[i + p - 1]] += 1
  answer += check()
}

console.log(answer)

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.py
import sys

input = sys.stdin.readline
s,p = map(int,input().split())
dna = list(input())
A,C,G,T = map(int,input().split())

currentChar = {
  'A':0,
  'C':0,
  'G':0,
  'T':0
}
def check():
  if currentChar['A'] >=A and currentChar['C'] >=C and currentChar['G'] >=G and currentChar['T'] >=T:
    return 1
  return 0

answer =0
for i in range(0,p):
  currentChar[dna[i]] +=1
answer += check()

for i in range(1,s-p+1):
  currentChar[dna[i-1]] -= 1
  currentChar[dna[i+p-1]] += 1
  answer +=check()

print(answer)

